### ISSUE TITLE
Precision Problem with Keras/Theano Sum()

### ISSUE THREAD

**Issue by mustgoplay** [fchollet/keras#5308]
There's a precision problem with keras/theano vs numpy.  

I have:
np.set_printoptions(precision=22)
K.set_epsilon(1e-20)

And am using either K.Sum() or T.Sum() vs np.Sum()

If I add these values:

[[  1.7117279052734375000000e+01,   1.7255817413330078125000e+01,
    1.7325582504272460937500e+01,   1.7325582504272460937500e+01,
    1.7255817413330078125000e+01,   2.2046510696411132812500e+01,
    2.2232553482055664062500e+01,   2.2325582504272460937500e+01,
    2.2325582504272460937500e+01,   2.2232553482055664062500e+01,
    2.4651163101196289062500e+01,   2.4883720397949218750000e+01,
    2.5000000000000000000000e+01,   2.5000000000000000000000e+01,
    2.4883720397949218750000e+01,   2.4651163101196289062500e+01,
    2.4883720397949218750000e+01,   2.5000000000000000000000e+01,
    2.5000000000000000000000e+01,   2.4883720397949218750000e+01,
    2.4651163101196289062500e+01,   2.4883720397949218750000e+01,
    2.5000000000000000000000e+01,   2.5000000000000000000000e+01,
    2.4883720397949218750000e+01]]
	
	In numpy I get:
	5.706986 7324829101562500e+02
	
	In Keras/theano I get:
	5.706986 6943359375000000e+02

	
	The problem is that as the calculations progress the level of error increases to the point that they're unusable.
	
Does anybody have any ideas please on how to get higher precision since setting a smaller epsilon didn't work?

Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x ] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [ ] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [ x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by patyork** [comment]
```python

orig = np.asarray([[  1.7117279052734375000000e+01,   1.7255817413330078125000e+01,
    1.7325582504272460937500e+01,   1.7325582504272460937500e+01,
    1.7255817413330078125000e+01,   2.2046510696411132812500e+01,
    2.2232553482055664062500e+01,   2.2325582504272460937500e+01,
    2.2325582504272460937500e+01,   2.2232553482055664062500e+01,
    2.4651163101196289062500e+01,   2.4883720397949218750000e+01,
    2.5000000000000000000000e+01,   2.5000000000000000000000e+01,
    2.4883720397949218750000e+01,   2.4651163101196289062500e+01,
    2.4883720397949218750000e+01,   2.5000000000000000000000e+01,
    2.5000000000000000000000e+01,   2.4883720397949218750000e+01,
    2.4651163101196289062500e+01,   2.4883720397949218750000e+01,
    2.5000000000000000000000e+01,   2.5000000000000000000000e+01,
    2.4883720397949218750000e+01]])

print '%.22f' % (np.sum(orig))
print '%.22f' % (np.sum(orig.astype('float32')))
```

yields
5.706986 7324829101562500e+02
5.706986 6943359375000000e+02

If you don't want to live in the limitations of foat32, you can switch to float64. However, few GPUs support double precision, so you'll be stuck with the CPU.

**Comment by the-moliver** [comment]
Yeah GPUs handle floating point math a little differently and the order in which things are summed in float32 can affect the result. In a sense, neither is really more correct than the other. Also if differences of .00001 matter for your application, you probably shouldn't be using neural nets, since 1) it's often a good idea to add huge amounts of noise during training (ie dropout), 2) retraining a network with different starting parameters will never give you the same answer to within .00001. So are you sure this is really a problem for you?

**Comment by mustgoplay** [comment]
Can you use float32 for most calcs and flloat64 for a subset?

It's definitely a problem because it's happening to each calculation in a series of calculations that include multiplications and divisions.  So if you have 2 similar additions at 2e5 the subtract them and divide by tgem the error can be .2-.3 which has a significant effect with weights.

**Comment by patyork** [comment]
You can do whatever you want, via the `cast` function, but more precision comes with a cost:

 - Just  doing this in a loss function may only affect running time by 1-5%, as the transferred data to the CPU is fairly small.
- Doing this in a custom layer may affect running time by 25-30% as the data transfer and subsequent CPU calculations will be done on quite a lot of data
  - Doing this poorly in a layer (forgetting to cast back to float32) will force all subsequent layers to be pushed the CPU as well, affecting runtime by 100+%

Also note: with your example above, float32 can't represent the sum of the array precisely. So keep in mind that if you do all of the calculations in float64 and cast the result down to float32, the final result will still most likely be off from what you were expecting:

```python
print '%.22f' % (np.sum(orig))
print '%.22f' % (np.sum(orig.astype('float32')))
print '%.22f' % np.asarray((np.sum(orig)), 'float32')
print '%.22f' % (np.asarray(570.6986732482910156250000, 'float32'))
```
```
570.6986 732482910156250000
570.6986 694335937500000000
570.6986 694335937500000000
570.6986 694335937500000000
```

**Comment by ahundt** [comment]
This problem is the purpose of the following algorithm:
https://en.wikipedia.org/wiki/Kahan_summation_algorithm

**Comment by patyork** [comment]
This is a pretty deep rabbit hole.. BLAS libs were something I'd always just accepted without reading about, so this was interesting.

Apparently the BLAS standards specifically avoid compensatory summation algs (like Kahan) for performance reasons, and I can't find any information on any compensatory algorithms in use by common BLAS libraries.

Theano and nvcc have flags that allow you to turn down precision on some ops (square root and division) to gain speed, but I don't think there's anything that deals specifically with summations there (no option to turn on compensatory algorithms). Python implements a compensatory algorithm (Shewchuk) that performs pretty good from what I can tell - not 100% sure it's actually doing it in float32 though.

Regardless, `570.6986 694335937500000000` is the closest float32 representable equivalent for `570.6986 732482910156250000`, so whatever operations actually go into the Theano sum are providing a value so close as to give back the closest representable number to the "actual" sum of the input (for this example, at least) - that's fairly impressive.

**Comment by ahundt** [comment]
> This is a pretty deep rabbit hole.

Definitely, haha. Adding floating point numbers in different orders produces different results in what is known as numerical error. https://sites.ualberta.ca/~kbeach/comp_phys/fp_err.html

@mustgoplay It is also important to note that according to my understanding, you simply won't always get the same exact result with these massively parallel algorithms because if you do large operations in parallel they often finish in varying orders, and may get added in the order of completion or split in varying ways, so the results vary, though it is usually very slightly.

**Comment by mustgoplay** [comment]
Thank you guys.

I'm doing simple, not hugely parallel, calcs but if I take the same exact image data and compare the normalized 0-1 to full 0-255 data, i.e. just multiply the normalized data by 255, I get completely different results.  Since I'm comparing data points, the results should be the same and they are the same in numpy.

This is really unfortunate because dealing with the MNIST training data it takes 26s in theano to do the equivalent, albeit inaccurate, calcs compared to 360+s in numpy.

Once I figure out how to force the calcs in theano (via cast or otherwise) to be 64-bit I'll see if there's any speed gain left from cpu optimizations.

I feel like I just wasted 2 weeks of my time converting the numpy code and having to figure out how to avoid loops in theano.

**Comment by ahundt** [comment]
which version of which backend are you using?

Looks like similar fixes may exist for tf releases?
> tf.round now uses banker's rounding (round to even) semantics to match NumPy.
from https://github.com/tensorflow/tensorflow/releases/tag/v1.0.0-rc2

**Comment by nouiz** [comment]
The new gpu back-end support float64, but to get decent speed, you need a
GPU with decent float64 speed. So no gamer card, no Titan card and only
some generation of scientific GPU have decent float64 speed.

Fred
